### DISCUSSION TITLE
Mount IDBFS in place of default MEMFS from pyscript

### DISCUSSION THREAD

**Discussion by kaclark** [pyscript/pyscript#1613] [Q&A] [answered]
I have not been able to figure out how to implement either the local file system or a persistent file system as of yet. During my search I came across pyodide documentation for this: https://pyodide.org/en/stable/usage/file-system.html

Is there a way to do something similar from purely pyscript?

Thanks

**Comment by JeffersGlass** [comment]
There sure is - you can check out the [demo and writeup I created](https://jeff.glass/post/whats-new-pyscript-2023-03-1/#native-filesystem) for PyScript's 2023.03.1 release.

It's not a fully-featured setup - minimal error handling, no fallback to something sensible etc., but it demonstrates the basic functionality.

**Reply by kaclark** [reply]
I am attempting to implement a website that can run in a mobile browser. The code provided on the website you've linked to seems to work well on desktop but throws an error on my mobile chrome. Would you know how to mount IDBFS in place of MEMFS? I'm noticing you've imported pyodide_js; is there a way I can install such a package such that something like "pyodide_js.FS.mount(pyodide.FS.filesystems.IDBFS, { root: "." }, mountDir);" would be valid inside pyscript tags?

**Reply by WebReflection** [reply]
I had issue in GNOME Web (Webkit based) ... are you having issues on iPhone (Safari, still WebKit based)???

It is possible they still don't allow read/write FileSystem API via browser, and if that's the case I am not sure we can do much to solve that, rather hope Safari on mobile will eventually provide such API.

**edit** P.S. also worth saying this functionality requires (AFAIK) https so you can't test it via local network addresses, unless you run on *localhost* but that's not usually the case for Mobile ... just sharing extra possible gotchas

**Reply by kaclark** [reply]
I was using google chrome on an android pixel; unsure how safari works here. GNOME on Fedora loaded up my file-browser, but I did not select anything to see if the full process worked.

**Reply by WebReflection** [reply]
do you see any error on Jeff's demo/link? I can see the same error in both GNOME Web and iPhone but it looks like a Pyodide issue with async listeners there ... as not available ... works for me in Chrome too ... just trying to understand what is it that you see (as issue).

**Reply by WebReflection** [reply]
![Screenshot from 2023-07-26 22-32-24](https://github.com/pyscript/pyscript/assets/85749/a32c17c2-ea72-4cd2-9cf0-5692639d837c)

**Reply by kaclark** [reply]
Jeff's demo works on my desktop. I get the same error on mobile. All this being said, it is not my real concern.

I'm mostly curious how I would interface with pyodide_js as shown in Jeff's example code. I cannot find any way to get that import to work. I am now attempting to implement IDBFS in place of the default MEMFS. 

I want to do this because I am looking for the functionality of saving a file with user input while hosted from a github page and to then load it from other sessions of the website. If my current trajectory is not fruitful, do you have any ideas for what could work to achieve this goal?

**Reply by WebReflection** [reply]
watch out IDBFS is based (if I'm not mistaken by its name) by IndexedDB which won't create a file in the phone/filesystem, it will persist arbitrarily (browser constraints) as data for that domain, so it's not like a real file like Jeff's demo would/could write/read from.

> I get the same error on mobile

that might be a Pyodide issue then, maybe trying with raw latest pyodide could bring you further. AFAIK we don't do anything special with that import for pyodide_js ... it's all part of pyodide API https://pyodide.org/en/stable/usage/api/python-api.html

**Reply by kaclark** [reply]
Data for the domain in fine for my use-case. How do I go about getting functionality access to the pyodide api? pip install did not work; what install process will make these resources available for use in python?

**Reply by WebReflection** [reply]
> Data for the domain in fine for my use-case

as long as you are aware iPhone might nuke that data after some week, that's good then!

> How do I go about getting functionality access to the pyodide api?

What I meant is that you might need to deal with raw pyodide, bootstrap it manually, forget about pyscript goodness, and see if with its latest version stuff works better ... after all, the code Jeff wrote simply uses pyscript classic utilities to evaluate some python you could evauate too with more orchestration, but it's not super magic, actually there's nothing in there that needs pyscript extra features for that specific tiny use case.

If you want me to write a whole example to try online, I am afraid you need to wait tomorrow as I am theoretically still on vacation until tomorrow :smile:

**Reply by kaclark** [reply]
Thank you so much for your continued support. I could happily wait a week for this example code. 

Above all else, what I am looking for is user input in a series of fields to be saved to a single file. I plan to then use the contents of this file to generate html. 

If there is the risk of the files being more or less deleted with time, then I could (though sub-optimal), maybe have a separate script set up to print out the raw contents of the text file and html and manually copy and paste those to files in the github dir.  I prefer complete automation, but this could work for a while.

**Reply by WebReflection** [reply]
let's see if I can manage to just reproduce Jeff's code in a way that uses latest pyodide and, if our browsers still show errors, I believe the next move would be to file a pyodide bug with such demo/example code and hope for a solution, if possible!

tomorrow dough :wave:

**Comment by kaclark** [comment]
Between Jeff's example code and the api you pointed me towards I was able to get the following working. I am not sure how to test this new file system, but I think I have mounted it. 

![image](https://github.com/pyscript/pyscript/assets/35940194/459041e9-e893-403d-8e30-a174f7027ec8)

**Reply by WebReflection** [reply]
OK, I put a test together https://webreflection.github.io/test/py-folder/

**edit** see [the answer](https://github.com/pyscript/pyscript/discussions/1613#discussioncomment-6563031)

If this does not work on mobile, we might want to go forward filing an issue in Pyodide.

**Reply by WebReflection** [reply]
**actually** ... the [showDirectoryPicker](https://developer.mozilla.org/en-US/docs/Web/API/Window/showDirectoryPicker#browser_compatibility) seems to be a Chrome Desktop only feature, as no other browser supports it ... I guess there's no bug to file in Pyodide, as this is the platform not providing the feature.

**Reply by WebReflection** [reply]
> I am not sure how to test this new file system, but I think I have mounted it.

if you can mount it, you can write in it too and, accordingly, you can then read it back ... give me a few minutes to write an example.

**Reply by WebReflection** [reply, marked as answer]
So ... this should actually work: https://webreflection.github.io/test/py-idbfs/ 

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Pyodide Persistent IDBFS</title>
  <script type="module">
    import plainTag from 'https://esm.sh/plain-tag';
    import { loadPyodide } from 'https://cdn.jsdelivr.net/pyodide/v0.23.4/full/pyodide.mjs';

    content.textContent = 'loading pyodide ...';
    const { FS, runPython } = await loadPyodide();
    content.textContent = '';

    const py = (...args) => runPython(plainTag(...args));

    // mount the IDBFS folder
    const mountDir = "/mnt";
    FS.mkdir(mountDir);
    FS.mount(FS.filesystems.IDBFS, { root: "." }, mountDir);
    FS.syncfs(true, error => {
      if (error) throw error;
      read();
    });
    py`import sys;sys.path.append("${mountDir}")`;

    const read = () => {
      py`
        from js import document
        import os
        import sys

        document.getElementById("content").textContent = "\\n".join(os.listdir("${mountDir}"))
      `;
    };

    write.disabled = false;
    write.addEventListener('click', event => {
      event.preventDefault();
      const content = prompt('What would you like to write?');
      if (content) {
        FS.writeFile(`${mountDir}/file_${+new Date}.txt`, content);
        FS.syncfs(error => {
          if (error) throw error;
        });
        read();
      }
    });
  </script>
</head>
<body>
  <button id="write" disabled>Write into folder</button>
  <pre id="content"></pre>
</body>
</html>
```

**Reply by WebReflection** [reply]
@hoodmane any idea why ~~above code won't be persistent?~~

**edit** fixed and updated the example.

**Reply by hoodmane** [reply]
Do you call `syncfs`?
https://emscripten.org/docs/api_reference/Filesystem-API.html#FS.syncfs

**Reply by WebReflection** [reply]
Ha! done that, it works now ... good catch!!!

So you have a whole example around read/write operations at this point. Happy to help forward but I think with the latest code you can also move forward too.

**Reply by kaclark** [reply]
Thank you for your help! I will test out what you have provided and then will attempt to integrate it into my current project.

Just before returning to this thread, I was poking around here:

https://emscripten.org/docs/api_reference/Filesystem-API.html#filesystem-api-idbfs

https://github.com/emscripten-core/emscripten/issues/15789

https://github.com/emscripten-core/emscripten/blob/main/test/fs/test_idbfs_sync.c

Your example is great! Persists across refresh; works on both my mobile browser and desktop.

**Reply by WebReflection** [reply]
You might want to flag that as an answer, for future readers, and it addresses all points in those topics:
  * the FS is read only after the initial (bootstrap) sync happens
  * every operation on write triggers a sync behind the scene ... IDB should be fast enough and the only issue I see is that a user closes the browsers little few ms after saving last file ... this is inevitable though, as `onbeforeunload` browser event no longer can really prevent users from closing a tab or their browser *but* I believe transitions triggered while closing on IDB side should still complete, as that's atomic and desired
  * it's true that having just a Promise based API there would avoid some boilerplate, but that's easy to implement anyway, so I think the current state is good enough

**Reply by kaclark** [reply]
My problem is completely resolved. 

I'm noticing that each device seems to store its own version, even when I access the html from a github static page. What documentation would be best for understanding this behavior? I mistakenly assumed that the static page hosting the website would hold the same files for all devices that visited it, and I'm curious as to why this is not the case and if it could ever be the case.

**Reply by WebReflection** [reply]
> I mistakenly assumed that the static page hosting the website would hold the same files for all devices that visited it

that would be a privacy violation, as any user could store private content then accessible to all others ... I am afraid IndexedDB works "like cookies" ... meaning it's per-user base.

To have a persistent FileSystem across users you need a backend that allows any visitor to mess up with such file system, and you can imagine that's a bad idea to start with for both users and the server itself, so that you should implement authentication per each user and be very defensive around the operations they perform.


> and I'm curious as to why this is not the case and if it could ever be the case.

I am afraid for the reasons I've mentioned and more, this will never be the case unless you pivot all operations to the server.

There are ways to do so, but it's out of PyScript scope and very risky for many security reasons.

**Reply by kaclark** [reply]
Thank you for this response. Everything you've said makes a great deal of sense. Once again, thank you for your help with everything!
